pallets/author/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, ensure, weights::Weight};
+use frame_system::{ensure_none, Config as System};
+use pallet_authorship::EventHandler;
+use parity_scale_codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use sp_inherents::ProvideInherentData;
+use sp_inherents::{InherentData, InherentIdentifier, IsFatalError, ProvideInherent};
+use sp_runtime::{ConsensusEngineId, DigestItem, RuntimeString};
+use sp_std::vec::Vec;
+
+pub trait Config: System {
+	/// Event type used by the runtime.
+	type Event: From<Event<Self>> + Into<<Self as System>::Event>;
+
+	/// Other pallets that want to be informed about block authorship.
+	/// We reuse the `pallet_authorship::EventHandler` for easy compatability with existing pallets.
+	type EventHandler: EventHandler<Self::AccountId, Self::BlockNumber>;
+}
+
+decl_event! {
+	pub enum Event<T> where
+		AccountId = <T as System>::AccountId,
+		BlockNumber = <T as System>::BlockNumber,
+	{
+		/// Set Author. Fields are Author, Block height
+		AuthorSet(AccountId, BlockNumber),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Config> {
+		/// Author already set in block.
+		AuthorAlreadySet,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as Author {
+		/// Author of current block.
+		Author: Option<T::AccountId>;
+
+		/// Did the author get set in this block?
+		DidUpdate: bool;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Inherent to set the author of a block
+		#[weight = 1_000_000]
+		fn set_author(origin, author: T::AccountId) {
+			ensure_none(origin)?;
+			ensure!(<Author<T>>::get().is_none(), Error::<T>::AuthorAlreadySet);
+
+			let current_block = frame_system::Module::<T>::block_number();
+
+			// Update storage
+			Author::<T>::put(&author);
+			DidUpdate::put(true);
+
+			// Add a digest item so Apps can detect the block author
+			// For now we use the Consensus digest item.
+			// Maybe this will change later.
+			frame_system::Module::<T>::deposit_log(DigestItem::<T::Hash>::Consensus(
+				ENGINE_ID,
+				author.encode(),
+			));
+
+			// Notify any other pallets that are listening (eg rewards) about the author
+			T::EventHandler::note_author(author.clone());
+
+			Self::deposit_event(Event::<T>::AuthorSet(author, current_block));
+		}
+
+		fn on_initialize() -> Weight {
+			// Reset the author at the beginning of the block
+			Author::<T>::kill();
+
+			// TODO how much weight should we actually be returning here.
+			0
+		}
+
+		fn on_finalize() {
+			// Ensure that the inherent is included in each block.
+			// This is how timestamp pallet does it.
+			// But there is also this provided method on the ProvideInherent trait.
+			// I wonder how it works
+			assert!(<Self as Store>::DidUpdate::take(), "Author inherent must be in the block");
```

Comment:
```suggestion
			assert!(<Author<T>>::get().is_some(), "Author inherent must be in the block");
```

Reply:
Interesting, it seems the way to ensure conditions are met in `on_finalize` is to panic if they're not met. This might be because nodes won't accept a block that panics in `on_finalize` so if they try it and it does panic, they reject it.

Reply:
That's my understanding too. The reason I used take here is because it also clears out the storage item as a side effect. (Copied that from timestamp)